### DISCUSSION TITLE
Camera Image Mask (circular)

### DISCUSSION THREAD

**Discussion by RoySalisbury** [aaronwmorris/indi-allsky#651] [Ideas]
There is a way to crop the image (x1, x2, y1, y2), but it would be nice if there was a way to select a circular region from the image (the fisheye lens area) and then mask out the rest of the image to reduce artifacts and bleed effects (bright lights at the edges).  

I have attached an image that shows the bleed and you can clearly see where the mask should be.

The final image is still rectangular, with all the metadata, but the ROI is just the non masked portion.  I've done this in my own software (C#) but have no idea how it would be done in python.

Roy  
![allskyimage](https://user-images.githubusercontent.com/12501943/226149636-a82bd649-c809-4d8c-a03f-db1d1915ef42.png)

**Comment by aaronwmorris** [comment]
Just yesterday I merged the ability to add a "Logo Overlay".  You can take a PNG with the same dimensions as your image and a alpha transparency layer can be layered on top of the image.

Here is an exaggerated example:
 
![cockpit](https://user-images.githubusercontent.com/17464290/226150341-3bb1cf07-4717-47cb-8b44-8df68c2b0325.jpg)

**Reply by RoySalisbury** [reply]
OK... Thats an interesting way to do it.  I'll give it a shot.

Thanks.

**Reply by StarGeezerPhil** [reply]
Works really well, I've just used it to tidy the image around the lens:

![Logo_Mask_Overlay](https://user-images.githubusercontent.com/94135145/226600916-299688cd-b216-4305-960d-a7959b76fe96.png)

**Reply by aaronwmorris** [reply]
Looks good.  I could have created a function to programmatically create a circle mask with a blurry edge, but there is a lot of variation between cameras.

For instance, this is an edge mask like yours... notice the image circle is bigger than the sensor and the image circle is offset from the center.

This could still have been done programmatically, but the configuration would have been more complex than I like to present to users.  Its easier to just take an image into Photoshop/Gimp and draw on top of it.

![asi385_logo3](https://user-images.githubusercontent.com/17464290/226651814-c65cacca-d100-4b7f-8760-9617e2afb70c.png)

**Comment by dbainbridge** [comment]
If user provided a circle radius or diameter value which is masked by the rectangular image size along with an x and y offset you could do it programmatically.  So you would just need 3 values and should be easily able to update values, verify fit, update values, etc.. until it fits properly.  Just an idea.

**Reply by aaronwmorris** [reply]
Of course, it is not impossible, but this logo mask kind of kills 3 birds with one stone.  StarGeezerPhil brought this up in https://github.com/aaronwmorris/indi-allsky/discussions/568

As implemented, it allows you to perform:

- Image cleanup
- Adding a logo
- Adding a privacy mask to blank out parts of the image

If everyone still thinks doing this programmatically is beneficial, we can always have both.

**Reply by RoySalisbury** [reply]
I think it would be beneficial.  But I am still going to try the original method you pointed out (the overlay).  I just have to find the right tools to do it with.  I don't do much 'editing' of images and the only software I have is PixInsight for my astronomy photos.

**Reply by aaronwmorris** [reply]
I will work on it.

**Reply by aaronwmorris** [reply]
Merged #658 to implement an image circle mask programmatically.

![Screenshot_20230322_170851](https://user-images.githubusercontent.com/17464290/227038907-c76d18fb-ba38-4617-9803-01a7859a8d12.png)

**Reply by RoySalisbury** [reply]
I was finally able to get a moment to update the software and try out the circle mask.

Perfect!

Here is a before and after shot...  
![before](https://user-images.githubusercontent.com/12501943/227745469-e555c3ea-7fb4-4193-b770-072ffd5b1725.png)

![after](https://user-images.githubusercontent.com/12501943/227745475-9091f518-6da4-4046-809c-454e3e75a91c.png)

**Reply by aaronwmorris** [reply]
Very nice!

**Comment by RoySalisbury** [comment]
I got the 'basic' idea working.  But its not feathered and its not exact .. but the best I could do wit the online tools I found...

![allsky_overlay](https://user-images.githubusercontent.com/12501943/226706793-386e479b-292e-40a2-b736-2a16e2ef1766.png)

**Reply by aaronwmorris** [reply]
It's pretty easy to do in Gimp (or Photoshop).  This is how I did it in Gimp.

1. Load image from camera, daytime is best
2. Create new layer, fill with foreground or background (which ever is black)
3. Reduce the opacity of the new black layer so that you can see the image in the background.  The opacity slider is just above the layer menu.
5. On the new layer, use the Ellipse Select Tool to draw your circle, drag edges to your liking.  Control-x to cut the hole.
7. Set the opacity of the layer back to 100%
8. Hide the background image (eye icon)
9. Filters -> Blur -> Gaussian Blur to smooth the edges
10. Export image as PNG
